cluster-autoscaler/clusterstate/clusterstate.go

```diff
@@ -111,6 +115,7 @@ type ScaleUpFailure struct {
 // ClusterStateRegistry is a structure to keep track the current state of the cluster.
 type ClusterStateRegistry struct {
 	sync.Mutex
+	context                            *context.AutoscalingContext
```

Comment:
I'm a bit sad we need to keep the entire context as a field here. Could we have an object that is just concerned with saying what max node provision time is for a given node (or nodegroup)? The context would then be kept there, making `ClusterStateRegistry` easier (well, at least not harder) to reason about.

Reply:
Renamed the provider interface to `maxNodeProvisionTimeProvider` and introduced `nodeRegistrationTimeLimitProvider` struct containing the provider and context. PTAL :)

Reply:
Thanks, although my point was really that whether or not the context is used is an implementation detail and doesn't have to be a part of the interface. So, one could imagine a `staticMaxNodeProvisionTimeProvider` which just returns a constant duration, as well as more advanced implementations depending on a specific use case. Regardless of what implementation of `maxNodeProvisionTimeProvider` interface is used, CSR should not require context passed to the constructor (i.e. `New...` function).

Reply:
Thank you for the explanation -  I've introduced `NewDefaultMaxNodeProvisionTimeProvider` returning an instance of `maxNodeProvisionTimeProvider`. It's kept as a field in `ClusterStateRegistry` and it already contains the context and `nodeGroupConfigProcessor` when it's assigned, so the `NewClusterStateRegistry` function signature is mostly back to as it was, it only takes the `maxNodeProvisionTimeProvider` as a new argument. Please take a look :)